core/spec/lib/spree/core/permalinks_spec.rb

```diff
@@ -0,0 +1,29 @@
+require 'spec_helper'
+
+describe Spree::Core::Permalinks do
+  describe "before_validation callbacks" do
+    let(:existing_permalink) { "ABCDEFG" }
+    let!(:existing_shipment) { create(:shipment, number: existing_permalink) }
+
+    let(:new_permalink)  { "123456789" }
+    let(:new_shipment) { build(:shipment) }
+
+    subject { new_shipment.save! }
+
+    context "generate_permalink returns an existing permalink" do
+      it "regenerates the permalink and saves to the model" do
+        expect(new_shipment).to receive(:generate_permalink).and_return(existing_permalink)
+        expect(new_shipment).to receive(:generate_permalink).and_return(new_permalink)
+
+        subject
+        expect(new_shipment.number).to eq(new_permalink)
+      end
+    end
+
+    context "generate_permalink returns a non-existing permalink" do
+      it "permalink is not nil" do
+        expect(subject).to_not be_nil
```

Comment:
subject.number gave me `undefined method "number" for true:TrueClass`,  instead I called subject first and then asserted on the new_shipment object. Let me know if you can think of a better way to handle that!